cupy/_core/dlpack.pyx

```diff
@@ -18,140 +13,246 @@ import warnings
+
+    device.device_id = array.data.device_id
+
     if not runtime._is_hip_environment:
-        attrs = runtime.pointerGetAttributes(data_ptr)
+        attrs = runtime.pointerGetAttributes(array.data.ptr)
         is_managed = (attrs.type == runtime.memoryTypeManaged)
         if is_managed:
             device.device_type = kDLCUDAManaged
-            dev_id = 0  # make it accessible on CPU too
         else:
             device.device_type = kDLCUDA
     else:
         device.device_type = kDLROCM
-    device.device_id = dev_id
-
-    cdef DLDataType* dtype = &dl_tensor.dtype
-    if array.dtype.kind == 'u':
-        dtype.code = <uint8_t>kDLUInt
-    elif array.dtype.kind == 'i':
-        dtype.code = <uint8_t>kDLInt
-    elif array.dtype.kind == 'f':
-        dtype.code = <uint8_t>kDLFloat
-    elif array.dtype.kind == 'c':
-        dtype.code = <uint8_t>kDLComplex
-    elif array.dtype.kind == 'b':
-        dtype.code = <uint8_t>kDLBool
+
+    return device
+
+
+# The name of this function is following the framework integration guide of
+# TensorComprehensions.
+cpdef object toDlpack(
+    _ndarray_base array, bint use_versioned=True, bint to_cpu=False,
+    bint ensure_copy=False, stream=None
+) except +:
+    """Create a dlpack capsule for an array.
+
+    Parameters
+    ----------
+    array : ndarray
+        The array to export
+    use_versioned : bool
+        Whether to use the versioned struct.  In the future this may be
+        which version to use.
+    to_cpu : bool
+        Whether we should make the data CPU available.
+    ensure_copy : bool
+        If `to_cpu` is True, whether a copy is requested/required.
+    stream : None or stream
+        Only used with `to_cpu`. The stream to use for making the data
+        available to the CPU.
+        If `None`, we make sure to synchronize to have the data available
+        as soon as we return.  Otherwise, we use this stream to copy the
+        data (as requested by the user).
+    """
+    cdef DLManagedTensor* dlm_tensor
+    cdef DLManagedTensorVersioned* dlm_tensor_ver
+    cdef DLTensor* dl_tensor
+    cdef const char *capsule_name
+
+    # Fetch dtype early (as this can raise a BufferError in theory)
+    cdef uint8_t dtype_code = get_dlpack_dtype_code(array.dtype)
+    cdef size_t dtype_itemsize = array.dtype.itemsize
+
+    # Fetch device information since we need it to deal with CPU logic.
+    cdef DLDevice device = get_dlpack_device(array)
+
+    cdef int32_t ndim = array._shape.size()
+
+    if not to_cpu:
+        owner = array
+    elif not ensure_copy and device.device_type == kDLCUDAManaged:
+        # Managed memory is CPU accessible.  Note that the consumer may expect
+        # `kDLCPU` here.  We only honor this request in spirit, but not
+        # strictly (because e.g. NumPy will remember the managed part).
+        owner = array
+        device.device_type = device.device_type
+        device.device_id = device.device_id
```

Comment:
```suggestion
```
Nothing, I previously changed it to `kDLCPU` here and didn't pay attention changing it.